diff --git a/engine/ggml/src/ggml-metal/ggml-metal-ops.cpp b/engine/ggml/src/ggml-metal/ggml-metal-ops.cpp
index c716f118..1e453a59 100644
--- a/engine/ggml/src/ggml-metal/ggml-metal-ops.cpp
+++ b/engine/ggml/src/ggml-metal/ggml-metal-ops.cpp
@@ -697,7 +697,9 @@ int ggml_metal_op_acc(ggml_metal_op_t ctx, int idx) {
 
         const int nth = std::min(ggml_metal_pipeline_max_theads_per_threadgroup(pipeline), ne00);
 
-        ggml_metal_encoder_dispatch_threadgroups(enc, ne01, ne02, ne03, nth, 1, 1);
+        const int nw0 = (ne00 + nth - 1)/nth;
+
+        ggml_metal_encoder_dispatch_threadgroups(enc, nw0*ne01, ne02, ne03, nth, 1, 1);
 
         ggml_metal_op_concurrency_reset(ctx);
     }
@@ -1790,7 +1792,9 @@ int ggml_metal_op_set(ggml_metal_op_t ctx, int idx) {
 
         const int nth = std::min(ggml_metal_pipeline_max_theads_per_threadgroup(pipeline), ne00);
 
-        ggml_metal_encoder_dispatch_threadgroups(enc, ne01, ne02, ne03, nth, 1, 1);
+        const int nw0 = (ne00 + nth - 1)/nth;
+
+        ggml_metal_encoder_dispatch_threadgroups(enc, nw0*ne01, ne02, ne03, nth, 1, 1);
 
         ggml_metal_op_concurrency_reset(ctx);
     }
